core/http-auth-spi/src/main/java/software/amazon/awssdk/http/auth/spi/AsyncHttpSignRequest.java

```diff
@@ -0,0 +1,59 @@
+/*
+ * Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License").
+ * You may not use this file except in compliance with the License.
+ * A copy of the License is located at
+ *
+ *  http://aws.amazon.com/apache2.0
+ *
+ * or in the "license" file accompanying this file. This file is distributed
+ * on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either
+ * express or implied. See the License for the specific language governing
+ * permissions and limitations under the License.
+ */
+
+package software.amazon.awssdk.http.auth.spi;
+
+import java.nio.ByteBuffer;
+import org.reactivestreams.Publisher;
+import software.amazon.awssdk.annotations.Immutable;
+import software.amazon.awssdk.annotations.SdkPublicApi;
+import software.amazon.awssdk.annotations.ThreadSafe;
+import software.amazon.awssdk.http.SdkHttpRequest;
+import software.amazon.awssdk.http.auth.spi.internal.DefaultAsyncHttpSignRequest;
+import software.amazon.awssdk.identity.spi.Identity;
+import software.amazon.awssdk.utils.builder.SdkBuilder;
+
+@SdkPublicApi
+@Immutable
+@ThreadSafe
+public interface AsyncHttpSignRequest<IdentityT extends Identity> extends HttpSignRequest<Publisher<ByteBuffer>, IdentityT> {
+    /**
+     * Get a new builder for creating a {@link AsyncHttpSignRequest}.
+     */
+    static <IdentityT extends Identity> Builder<IdentityT> builder(Class<IdentityT> ignoredIdentityType) {
+        return new DefaultAsyncHttpSignRequest.BuilderImpl<>();
+    }
+
+    @Override
+    default Class<Publisher<ByteBuffer>> payloadType() {
+        // TODO: Code Review Note: Note this cast
+        return (Class) Publisher.class;
+    }
```

Comment:
Do we still need this getter? We could split SignedHttpRequest like we split HttpSignRequest into sync/async so that we don't need this anymore.

Reply:
Yeah, tradeoff would be Signer implementations where the logic for sync/async isn't different, will have some code duplication, see HttpSignerTest in this PR. Probably applies to bearer, apiKey, httpBasic/Digest. So I thought this one cast was an okay tradeoff. LMK if you prefer otherwise. Or if you think, if we are splitting the input, the output should also be split anyways, or split neither?

The alternate, without splitting either would be to remove the `Class<T>` payload parameter in the builder() method (and the `payloadType()` accessor). The drawback being that someone using `HttpSignRequest.builder()` would be wondering why they are getting `Object` for `payload()`. I think this might be the better tradeoff, as the compiler helps the customer there. The use of these types is for the sign methods and they expect specific types, so it will be easy enough to understand that `<Publisher<ByteBuffer>>` needed to be added. Moreover, this is ony needed if using the builder directly. We can add the Consumer Builder methods still to alleviate the need to instantiate the builder directly.

Reply:
Resolved by moving to separate sub types for sync and async for SignedHttpRequest too and removing the Class<T> when creating the builder as well as returning it.